hibernate-core/src/main/java/org/hibernate/Hibernate.java

```diff
@@ -239,4 +259,158 @@ public static Object unproxy(Object proxy) {
 	public static <T> T unproxy(T proxy, Class<T> entityClass) {
 		return entityClass.cast( unproxy( proxy ) );
 	}
+
+	/**
+	 * Obtain a detached, uninitialized reference (a proxy) for a persistent entity with the given identifier.
+	 *
+	 * The returned proxy is not associated with any session, and cannot be initialized by calling
+	 * {@link #initialize(Object)}. It can be used to represent a reference to the entity when working with
+	 * a detached object graph.
+	 *
+	 * @param sessionFactory the session factory with which the entity is associated
+	 * @param entityClass the entity class
+	 * @param id the id of the persistent entity instance
+	 *
+	 * @return a detached uninitialized proxy
+	 */
+	public static <E> E createDetachedProxy(SessionFactory sessionFactory, Class<E> entityClass, Object id) {
+		EntityPersister persister =
+				sessionFactory.unwrap(SessionFactoryImplementor.class).getMetamodel()
+						.findEntityDescriptor(entityClass);
+		if (persister==null) {
+			throw new UnknownEntityTypeException("unknown entity type");
+		}
+		return (E) persister.createProxy(id, null);
+	}
+
+	/**
+	 * Operations for obtaining references to persistent collections of a certain type.
+	 *
+	 * @param <C> the type of collection, for example, <tt>List&lt;User&gt;</tt>
+	 */
+	public static final class CollectionInterface<C> {
```

Comment:
I don't see why it is useful to have this i.e. it doesn't do anything a user would not be able to do through `new WhateverCollection()` or `new PersistentWhateverCollection()`. Creating a collection proxy on the other hand would IMO be something new that would have merit to have, but like I stated before, I'd rather like an API like this:

```
public static <C> C createProxy(PluralAttribute<C, ?, ?> attr) {...}
```

Reply:
> I don't see why it is useful to have this

I agree it's not useful for people handwriting code, but that's not the main usecause here. It might be quite useful for generic code, i.e. serialization frameworks, which is how this is most likely to be used.

> I'd rather like an API like this:

I have absolutely no problem with providing an API like that, and indeed it can be easily implemented on top of what I've proposed. But the API you just showed is _much_ less convenient for framework code, and the extra typesafety is completely useless, since the framework code is going to be working with things in a detyped way.

Summary: I've designed this with the needs of a serialization framework developer in mind.

Reply:
But even the serialization framework needs some sort of type structure contract to traverse an object along, and why not use the JPA metmodel for this purpose? Apart from that, I would also add 

```
public static <V> V createDefault(Attribute<?, V> attr) {...}
```

To create whatever default value should be used for an attribute which could be an `ArrayList`, `null` or some primitive value.

Reply:
The serialization framework might not even have access to a `SessionFactory`.

Reply:
And, I mean, it's almost  certainly *not* going to be traversing the Hibernate meta model! It's going to have its whole own logic for walking an object graph. 

The user is just plugging in custom logic for dealing with Hibernate proxies.

Reply:
I don't know, I would expect the `SessionFactory` to be available during deserialization, but I know there are some serialization framworks that make this really hard, so yeah, maybe this has merit. IMO we should have at least 1 user of this API (even if it is just us trying it out in some deserializer) that confirms this works great for some purpose before we introduce it. People that write this custom logic could just as well write `new PersistentWhateverCollection()`, so I'm questioning what the particular value of this code is. Is it just convenience?

Reply:
The value is you're not telling people to call what is quite clearly marked as an internal API that their program is not supposed to depend on?

Reply:
Hmm, ok, that makes sense. I didn't realize the persistent collections are internal API.

Reply:
> Hmm, ok, that makes sense. I didn't realize the persistent collections are internal API.

We really should have a review of that.  Conceptually I'd say it is a mix - some of it could easily be API/SPI while other parts are clearly internal details.

But I think the PR is fine, as far as it goes.

Reply:
> We really should have a review of that. Conceptually I'd say it is a mix - some of it could easily be API/SPI while other parts are clearly internal details.

That would be worth doing indeed.